src/users.rs

```
@@ -0,0 +1,243 @@
+    salt
+}
+
+// Takes in a salt and password and returns the hash
+// using the pbkdf2 algorithm
+fn get_hash_from_password(password: String, salt: Vec<u8>) -> Vec<u8> {
+    let mut pbkdf2_hash = [0u8; CREDENTIAL_LEN];
+    pbkdf2::derive(
+        &digest::SHA512,
+        N_ITER,
+        &salt,
+        password.as_bytes(),
+        &mut pbkdf2_hash,
+    );
+    pbkdf2_hash.to_vec()
+}
+
+// Verifies a password and salt against a hash
+// using the pbkdf2 algorithm
+fn verify_password(
+    password: String,
+    salt: Vec<u8>,
+    pbkdf2_hash: Vec<u8>,
+) -> Result<(), Unspecified> {
+    pbkdf2::verify(
+        &digest::SHA512,
+        N_ITER,
+        &salt,
+        password.as_bytes(),
+        &pbkdf2_hash,
+    )
+}
+
+// creates a sqlite DB if it does not exist and initializes the user table
+// returns the connection
+fn connection() -> Result<rusqlite::Connection, Box<dyn Error>> {
+    let conn = Connection::open("sqlite.db").expect("db conn fail");
+
+    // create the users table if it does not exist
+    match conn.execute(
+        "create table if not exists users (
+             id integer primary key,
+             name text not null unique,
+             salt text not null,
+             hash text not null,
+             active integer default 1 not null
+         );",
+        NO_PARAMS,
+    ) {
+        Ok(_) => {
+            return Ok(conn);
+        }
+        Err(_) => {
+            return Err(Box::new(UserError("Error Creating table".into())));
+        }
+    };
+}
+
+impl User {
+    // trait used to create a user
+    pub fn create(&mut self, password: String) -> Result<(), Box<dyn Error>> {
+        // Generates a salt for the user
+        let salt = get_salt();
+        self.salt = HEXUPPER.encode(&salt);
+        // Generates a hash from the salt and password
+        let hash = get_hash_from_password(password, salt);
+        self.hash = HEXUPPER.encode(&hash);
+        // Connect to the db
+        match connection() {
```

Same as above regarding error handling - can we just return the original errors?